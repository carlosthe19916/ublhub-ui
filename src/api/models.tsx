export interface PageQuery {
  page: number;
  perPage: number;
}

export interface SortByQuery {
  orderBy: string | undefined;
  orderDirection: "asc" | "desc";
}

export interface PageRepresentation<T> {
  meta: Meta;
  links: Links;
  data: T[];
}

export interface Meta {
  offset: number;
  limit: number;
  count: number;
}

export interface Links {
  first: string;
  next: string;
  previous: string;
  last: string;
}
export interface Company {
  id?: string;
  name: string;
  description?: string;
  sunatWsUrls: SUNATWsUrls;
  sunatCredentials: SUNATCredentials;
}

export interface SUNATWsUrls {
  factura: string;
  guia: string;
  retencion: string;
}

export interface SUNATCredentials {
  username: string;
  password: string;
}
