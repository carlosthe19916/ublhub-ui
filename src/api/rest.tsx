import { AxiosPromise } from "axios";
import { APIClient } from "axios-config";

import { Company, PageQuery, PageRepresentation, SortByQuery } from "./models";

const USER_COMPANIES = "/user/companies";
const COMPANIES = "/companies";

export const DOCUMENTS = `${COMPANIES}/:company/documents`;

export const createCompany = (company: any): AxiosPromise<Company> => {
  return APIClient.post(USER_COMPANIES, company);
};

export const getCompanies = (
  pagination: PageQuery,
  sortBy?: SortByQuery,
  filterText?: string
): AxiosPromise<PageRepresentation<Company>> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    sortByQuery = `${sortBy.orderBy}:${sortBy.orderDirection}`;
  }

  const params = {
    filterText,
    offset: (pagination.page - 1) * pagination.perPage,
    limit: pagination.perPage,
    sort_by: sortByQuery,
  };

  const query: string[] = [];
  Object.keys(params).forEach((key) => {
    const value = (params as any)[key];
    if (value !== undefined) {
      query.push(`${key}=${value}`);
    }
  });

  return APIClient.get(`${USER_COMPANIES}?${query.join("&")}`);
};

export const deleteCompany = (company: Company): AxiosPromise => {
  return APIClient.delete(`${COMPANIES}/${company.name}`);
};

export const getCompany = (name: string): AxiosPromise<Company> => {
  return APIClient.get(`${COMPANIES}/${name}`);
};
