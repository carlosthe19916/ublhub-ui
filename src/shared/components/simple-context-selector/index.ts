export { SimpleContextSelector } from "./simple-context-selector";
