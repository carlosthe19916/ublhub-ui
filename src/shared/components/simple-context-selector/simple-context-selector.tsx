import React, { useState } from "react";
import { ContextSelector, ContextSelectorItem } from "@patternfly/react-core";
import { Namespace } from "api/models";

const filterItems = (items: Namespace[], filterText: string) => {
  const filtered =
    filterText === ""
      ? [...items]
      : items.filter(
          (f) => f.name.toLowerCase().indexOf(filterText.toLowerCase()) !== -1
        );

  return filtered;
};

export interface SimpleContextSelectorProps {
  value?: Namespace;
  items: Namespace[];
  onChange: (value: Namespace) => void;
}

export const SimpleContextSelector: React.FC<SimpleContextSelectorProps> = ({
  value,
  items,
  onChange,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [searchValue, setSearchValue] = useState("");

  const handleOnSelect = (_: any, value: any) => {
    const selectedItem = items.find((f) => f.name === value);
    setIsOpen((current) => !current);
    if (selectedItem) {
      onChange(selectedItem);
    }
  };

  const handleOnToggle = (_: any, isOpen: boolean) => {
    setIsOpen(isOpen);
  };

  const handleOnSearchInputChange = (value: string) => {
    setSearchValue(value);
  };

  return (
    <ContextSelector
      toggleText={value?.name}
      onSearchInputChange={handleOnSearchInputChange}
      isOpen={isOpen}
      searchInputValue={searchValue}
      onToggle={handleOnToggle}
      onSelect={handleOnSelect}
      screenReaderLabel="Seleccione:"
    >
      {filterItems(items, searchValue).map((item, index) => (
        <ContextSelectorItem key={index}>{item.name}</ContextSelectorItem>
      ))}
    </ContextSelector>
  );
};
