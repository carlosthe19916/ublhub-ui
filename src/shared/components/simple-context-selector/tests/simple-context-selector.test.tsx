import React from "react";
import { mount, shallow } from "enzyme";
import { SimpleContextSelector } from "../simple-context-selector";
import { Namespace } from "api/models";

describe("SimpleContextSelector", () => {
  const namespaces: Namespace[] = [
    { name: "ns1" },
    { name: "ns2" },
    { name: "ns3" },
  ];

  it("Renders without crashing", () => {
    const wrapper = shallow(
      <SimpleContextSelector
        value={namespaces[0]}
        items={namespaces}
        onChange={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Select item", () => {
    const onChangeSpy = jest.fn();

    const wrapper = mount(
      <SimpleContextSelector items={namespaces} onChange={onChangeSpy} />
    );

    wrapper.find(".pf-c-context-selector__toggle").simulate("click");
    wrapper
      .find(".pf-c-context-selector__menu-list-item")
      .at(1)
      .simulate("click");

    expect(onChangeSpy).toHaveBeenCalledTimes(1);
    expect(onChangeSpy).toBeCalledWith(namespaces[1]);
  });

  it("Filter items", () => {
    const onChangeSpy = jest.fn();

    const wrapper = mount(
      <SimpleContextSelector items={namespaces} onChange={onChangeSpy} />
    );

    wrapper.find(".pf-c-context-selector__toggle").simulate("click");

    const searchInput = wrapper.find("input");
    (searchInput.getDOMNode() as HTMLInputElement).value = "1";
    searchInput.simulate("change");
    wrapper.update();

    expect(wrapper.find(".pf-c-context-selector__menu-list-item").length).toBe(
      1
    );
    expect(wrapper.find(".pf-c-context-selector__menu-list-item").text()).toBe(
      "ns1"
    );
  });
});
