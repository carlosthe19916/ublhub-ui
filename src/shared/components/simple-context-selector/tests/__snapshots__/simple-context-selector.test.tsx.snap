// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleContextSelector Renders without crashing 1`] = `
<ContextSelector
  className=""
  footer={null}
  isOpen={false}
  menuAppendTo="inline"
  onSearchButtonClick={[Function]}
  onSearchInputChange={[Function]}
  onSelect={[Function]}
  onToggle={[Function]}
  ouiaSafe={true}
  screenReaderLabel="Seleccione:"
  searchButtonAriaLabel="Search menu items"
  searchInputPlaceholder="Search"
  searchInputValue=""
  toggleText="ns1"
>
  <ContextSelectorItem
    className=""
    isDisabled={false}
    key="0"
    onClick={[Function]}
    sendRef={[Function]}
  >
    ns1
  </ContextSelectorItem>
  <ContextSelectorItem
    className=""
    isDisabled={false}
    key="1"
    onClick={[Function]}
    sendRef={[Function]}
  >
    ns2
  </ContextSelectorItem>
  <ContextSelectorItem
    className=""
    isDisabled={false}
    key="2"
    onClick={[Function]}
    sendRef={[Function]}
  >
    ns3
  </ContextSelectorItem>
</ContextSelector>
`;
